### PR TITLE
HSEARCH-2393 Remove dependency to Lucene in tests for features common to Lucene and Elasticsearch

### PR DESCRIPTION
**Pull request by yrodiere**
This PR is based on HSEARCH-2296, we'll have to review and merge the PR ( #1192) first.

Related JIRA ticket: https://hibernate.atlassian.net/browse/HSEARCH-2393

The purpose of this PR is to make tests currently ignored in `elasticsearch/pom.xml` work with Elasticsearch, so that we can enable them.

### CODE REVIEW

File: orm/src/test/java/org/hibernate/search/test/interceptor/Blog.java [hunk 1]
```diff
@@ -29,6 +31,18 @@ public void setId(Integer id) {
 		this.id = id;
 	}
 
+	@Override
+	public boolean equals(Object obj) {
+		return getClass().equals( obj.getClass() )
+				&& id != null
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/interceptor/Blog.java]: Not worth a fix but we agree we don't need to test for the id to be not null here?

File: orm/src/test/java/org/hibernate/search/test/interceptor/Blog.java [hunk 2]
```diff
@@ -29,6 +31,18 @@ public void setId(Integer id) {
...
+				&& Objects.equals( id, ((Blog) obj).id );
+	}
+
+	@Override
+	public int hashCode() {
+		return getClass().hashCode();
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/interceptor/Blog.java]: I'm surprised the `hashCode` only use the class `hashCode`, while the `equals` involves the id?

File: orm/src/test/java/org/hibernate/search/test/interceptor/IndexingActionInterceptorTest.java [hunk 3]
```diff
@@ -64,17 +62,18 @@ public void tearDown() throws Exception {
 	public void testBlogAndArticleAreNotIndexedInDraftStatus() throws Exception {
 		Transaction tx = fullTextSession.beginTransaction();
 
-		assertThat( getBlogEntriesFor( Blog.class ) ).as( "Blog is explicit intercepted" ).hasSize( 0 );
-		assertThat( getBlogEntriesFor( Article.class ) ).as( "Article is inherently intercepted" ).hasSize( 0 );
+		List<Blog> blogEntries = getBlogEntries();
+		assertThat( blogEntries ).as( "Blog is explicit intercepted" ).excludes( blog );
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/interceptor/IndexingActionInterceptorTest.java]: If we end up amending this commit, let's fix the message and use "explicitly".

File: orm/src/test/java/org/hibernate/search/test/interceptor/IndexingActionInterceptorTest.java [hunk 4]
```diff
@@ -64,17 +62,18 @@ public void tearDown() throws Exception {
...
 		tx.commit();
 	}
 
 	@Test
 	public void testTotalArticleIsIndexedInDraftStatus() throws Exception {
 		Transaction tx = fullTextSession.beginTransaction();
-		assertThat( getBlogEntriesFor( TotalArticle.class ) ).as( "TotalArticle is explicit not intercepted" )
-				.hasSize( 1 );
+		List<Blog> blogEntries = getBlogEntries();
+		assertThat( blogEntries ).as( "TotalArticle is explicit not intercepted" ).contains( totalArticle );
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/interceptor/IndexingActionInterceptorTest.java]: Same here.

File: orm/src/test/java/org/hibernate/search/test/interceptor/IndexingActionInterceptorTest.java [hunk 5]
```diff
@@ -109,8 +109,8 @@ public void testTotalArticleIsIndexedInRemovedStatus() throws Exception {
 		setAllBlogEntriesToStatus( BlogStatus.REMOVED );
 		Transaction tx = fullTextSession.beginTransaction();
 
-		assertThat( getBlogEntriesFor( TotalArticle.class ) ).as( "TotalArticle is explicit not intercepted" )
-				.hasSize( 1 );
+		List<Blog> blogEntries = getBlogEntries();
+		assertThat( blogEntries ).as( "TotalArticle is explicit not intercepted" ).contains( totalArticle );
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/interceptor/IndexingActionInterceptorTest.java]: Well, let's do a search and replace if we decide to fix it :).

File: orm/src/test/java/org/hibernate/search/test/SearchInitializationTestBase.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * <p>The main difference with {@link SearchTestBase} is that the test itself does not implement
+ * {@link TestConfiguration}, which enables using different configurations for different test methods.
+ * <p>When subclassing this class, Hibernate Search's initialization must be triggered manually by calling
+ * {@link #init(TestConfiguration)}.
+ * <p>This is most commonly used to assert exceptions are thrown in specific cases.
+ *
+ * @author Yoann Rodiere
+ */
+//@RunWith(CustomRunner.class)
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/SearchInitializationTestBase.java]: Maybe remove the commented `@RunWith`?

File: orm/src/test/java/org/hibernate/search/test/SearchInitializationTestBase.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import org.hibernate.search.test.util.ImmutableTestConfiguration;
+import org.hibernate.search.test.util.TestConfiguration;
+import org.junit.After;
+
+/**
+ * A base class for tests of the initialization process.
+ *
+ * <p>The main difference with {@link SearchTestBase} is that the test itself does not implement
+ * {@link TestConfiguration}, which enables using different configurations for different test methods.
+ * <p>When subclassing this class, Hibernate Search's initialization must be triggered manually by calling
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/SearchInitializationTestBase.java]: If we end up amending this commit, I don't think the `'s` is in order here. No need to fix it if we don't amend the commit though.

File: orm/src/test/java/org/hibernate/search/test/SearchInitializationTestBase.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	 * Initialize with no particular settings
+	 * @param annotatedClasses The classes to search for annotations
+	 */
+	protected void init(Class<?>... annotatedClasses) {
+		Map<String, Object> settings = new HashMap<>();
+		init( new ImmutableTestConfiguration( settings, annotatedClasses ) );
+	}
+
+	/**
+	 * @return the testResourceManager, or null there isn't any (i.e. if {@link #init(TestConfiguration)}
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/SearchInitializationTestBase.java]: Missing "if" maybe?

File: engine/src/test/java/org/hibernate/search/test/util/impl/ExpectedLog4jLog.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+	public static ExpectedLog4jLog create() {
+		return new ExpectedLog4jLog();
+	}
+
+	private List<Matcher<?>> expectations = new ArrayList<>();
+
+	private ExpectedLog4jLog() {
+	}
+
+	public Statement apply(Statement base, org.junit.runner.Description description) {
```
Comment (gsmet) [inline engine/src/test/java/org/hibernate/search/test/util/impl/ExpectedLog4jLog.java]: Missing `@Override`?

File: documentation/src/main/asciidoc/elasticsearch-integration.asciidoc [hunk 10]
```diff
@@ -173,6 +173,10 @@ In order for your value to be understood by Hibernate Search (and Elasticsearch)
  * For booleans, use either `true` or `false`.
  * For dates, use the ISO-8601 format (`yyyy-MM-dd'T'HH:mm:ssZ`, for instance `2016-08-26T16:41:00+01:00`). The time and time zone may be omitted (if omitted, the time zone will be interpreted as the default JVM time zone).
 
+===== DynamicBoost
```
Comment (gsmet) [inline documentation/src/main/asciidoc/elasticsearch-integration.asciidoc]: Maybe "Dynamic boosting" would be a better title?

File: documentation/src/main/asciidoc/elasticsearch-integration.asciidoc [hunk 11]
```diff
@@ -173,6 +173,10 @@ In order for your value to be understood by Hibernate Search (and Elasticsearch)
...
+
+The `org.hibernate.search.annotations.DynamicBoost` is not (and cannot be) supported with Elasticsearch, because the platform lacks per-document, index-time boosting capabilities. Static boost (`@Boost`) is, however, supported.
```
Comment (gsmet) [inline documentation/src/main/asciidoc/elasticsearch-integration.asciidoc]: I would have use the plural form:
> Static boosts (`@Boost`) are, however, supported.

I think it sounds more natural but maybe it's just me.

File: elasticsearch/src/test/java/org/hibernate/search/elasticsearch/test/ElasticsearchUnsupportedIT.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+ * @author Yoann Rodiere
+ */
+public class ElasticsearchUnsupportedIT extends SearchInitializationTestBase {
+
+	@Rule
+	public ExpectedLog4jLog logged = ExpectedLog4jLog.create();
+
+	@Test
+	public void dynamicBoosting() throws Exception {
+		logged.expectMessage( "HSEARCH400032", "@DynamicBoost", DynamicBoostingEntity.class.getSimpleName() );
```
Comment (gsmet) [inline elasticsearch/src/test/java/org/hibernate/search/elasticsearch/test/ElasticsearchUnsupportedIT.java]: Nice trick!

File: elasticsearch/src/test/java/org/hibernate/search/elasticsearch/test/ElasticsearchUnsupportedIT.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+import org.hibernate.search.test.util.impl.ExpectedLog4jLog;
+import org.junit.Rule;
+import org.junit.Test;
+
+/**
+ * Test that users attempting to use unsupported features are duly warned.
+ *
+ * @author Yoann Rodiere
+ */
+public class ElasticsearchUnsupportedIT extends SearchInitializationTestBase {
```
Comment (gsmet) [inline elasticsearch/src/test/java/org/hibernate/search/elasticsearch/test/ElasticsearchUnsupportedIT.java]: Maybe ElasticsearchUnsupported**Features**IT?

File: orm/src/test/java/org/hibernate/search/test/jpa/ToStringTest.java [hunk 14]
```diff
@@ -66,16 +67,16 @@ public void testToStringContainsQueryInformationForJPAUseCase() throws Exception
...
-		assertEquals(
-				"Unexpected toString implementation. The string should contain Lucene query information",
-				"FullTextQueryImpl(fubar:snafu)",
-				fullTextQueryString
-		);
+	private void assertToStringContainsLuceneQueryInformation(String fullTextQueryToString) {
+		String queryString = ContextHelper.getSearchIntegrator( fullTextSession )
+				.createHSQuery( luceneQuery, Foo.class )
+				.getQueryString();
+		assertThat( "Unexpected toString implementation. The string should contain Lucene query information",
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/jpa/ToStringTest.java]: Very minor: is it a Lucene query even when using Elasticsearch?

File: elasticsearch/pom.xml [hunk 15]
```diff
@@ -234,7 +234,6 @@
                                 
                                 <!-- HSEARCH-2393 Remove depencies to Lucene in tests for features common to Lucene and Elasticsearch -->
                                 <exclude>**/ProgrammaticMappingTest.java</exclude><!-- Analyzers -->
-                                <exclude>**/ProjectionQueryTest.java</exclude><!-- Projection constant __HSearch_Document -->
```
Comment (gsmet) [inline elasticsearch/pom.xml]: Not related to this line but to the commit message of 78a0f9add77d9180aacd309f3c737237a1074f64: look for "HSEARCH-HSEARCH"

File: orm/src/test/java/org/hibernate/search/test/engine/NumericFieldTest.java [hunk 16]
```diff
@@ -205,29 +211,29 @@ public void testByteDocumentIdMappedAsStringFieldByDefault() {
...
-			List<Object[]> list = fullTextSession.createFullTextQuery( query, TouristAttraction.class )
-					.setProjection( ProjectionConstants.DOCUMENT )
-					.list();
-
-			assertEquals( 1, list.size() );
-			Document document = (Document) list.iterator().next()[0];
-
-			IndexableField scoreNumeric = document.getField( "scoreNumeric" );
-			assertThat( scoreNumeric.numericValue() ).isEqualTo( 23 );
+		assertEquals(
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/engine/NumericFieldTest.java]: I am not sure about removing these tests. Do we test elsewhere that the indexed fields are effectively numeric? If not, I would add your new test for Es/Lucene agnostic tests that the field bridge is OK but I think I would create backend specific tests to check that the indexed document is OK.

File: orm/src/test/java/org/hibernate/search/test/query/sorting/Explorer.java [hunk 17]
```diff
@@ -50,9 +50,12 @@
...
 	@ManyToOne
 	@IndexedEmbedded
-	@Field(bridge = @FieldBridge(impl = Territory.IdFieldBridge.class))
-	@NumericField
-	@SortableField
+	/*
+	 * Don't store the id in "favoriteTerritory" directly as this could conflict
+	 * with the @IndexedEmbedded with the Elasticsearch backend
+	 */
+	@Field(name = "favoriteTerritory.idFromBridge", bridge = @FieldBridge(impl = Territory.IdFieldBridge.class))
```
Comment (gsmet) [inline orm/src/test/java/org/hibernate/search/test/query/sorting/Explorer.java]: Interesting. In which case might we have a conflict? I suppose it's related to the null case (just a vague remembering of a previous PR)? Maybe we should add a prefix or a suffix for this field specific to Es to avoid conflicts?

File: orm/src/test/java/org/hibernate/search/test/configuration/ProgrammaticMappingTest.java [hunk 18]
```diff
@@ -103,41 +107,30 @@ public void testMapping() throws Exception {
...
-		Object[] row = result.iterator().next();
-		Document document = (Document) row[0];
-
-		IndexableField priceNumeric = document.getField( "price" );
-		assertThat( priceNumeric.numericValue() ).isEqualTo( 3454 );
-
-		IndexableField priceString = document.getField( "price_string" );
-		assertThat( priceString.numericValue() ).isNull();
-		assertThat( priceString.stringValue() ).isEqualTo( "3454" );
+		assertEquals(
```
Comment (yrodiere) [inline orm/src/test/java/org/hibernate/search/test/configuration/ProgrammaticMappingTest.java]: And same answer :) I created `LuceneProgrammaticMappingTest` nevertheless.

**Comment by gsmet** [comment]
Yes, it's good to go!
